specs/agents/tracing-spans-service-target.md

```diff
@@ -0,0 +1,236 @@
+  }
+  
+  if (!service_target.name) { // infer name from span attributes
+      
+    if (span.context.db.instance) {  // database spans
+      service_target.name = span.context.db.instance;
+
+    } else if (span.context.message) { // messaging spans
+      service_target.name = span.context.message.queue.name
+
+    } else if (context.http.url) { // http spans
+      service_target.name = context.http.host;
+      if (context.http.url.port > 0) {
+        service_target.name += ":" + context.http.port;
+      }
+    }
+
+  }
+} else {
+    // non-exit spans should not have service.target.* fields
+    span.context.service.target = undefined;
+}
+```
+
+The values for `span.context.db.instance` are described in [SQL Databases](./tracing-instrumentation-db.md#sql-databases).
+
+The values for `span.context.message.queue.name` are described in [Messaging context fields](./tracing-instrumentation-messaging.md#context-fields)
+
+### User API
+
+Agents SHOULD provide an API entrypoint to set the value of `span.context.destination.service.resource`,
+setting an empty or `null` value allows the user to discard the inferred value.
+This API entrypoint should be marked as deprecated and replaced by the following:
+
+Agents SHOULD provide an API entrypoint to set the value of `span.context.service.target.type` and `span.context.service.target.name`,
+setting an empty or `null` value on both of those fields allows the user to discard the inferred values.
+
+When a user-provided value is set, it should take precedence over inferred values from the span `_.type` `_.subtype` or any `_.context` attribute.
+
+### Phase 3
+
+In Phase 3 the `span.destination.service.resource` field value is inferred on APM Server with the following algorithm:
+
+```javascript
+
+// span from agent intake
+span = {};
+
+// inferred destination resource (if any)
+destination_resource = undefined;
+
+if (span.service.target.type) {
+  // new fields, store them as-provided
+  // infer resource destination from new fields
+  destination_resource = span.service.target.type;
+  if (span.service.target.name) {
+    destination_resource += "/";
+    destination_resource += span.service.target.name;
+  }
+
+} else if (['db', 'messaging', 'external'].indexOf(span.type) >= 0) {
+  // agent intake without new fields
+  // only a subset of span types should be included
+  span.service.target = {};
+
+  // inferred service target type & name
+  target_type = undefined;
+  target_name = undefined;
+
+  if (span.destination.service.resource) {
+    // infer new fields values from destination resource if provided
+    r = span.destination.service.resource;
+    separatorIndex = r.indexOf('/');
+    if (separatorIndex <= 0) {
+      target_type = r;
+    } else {
+      target_type = r.substr(0, separatorIndex);
+      target_name = r.substr(separatorIndex + 1);
+    }
+  } else {
+    // no destination resource, fallback from span type & subtype
+    target_type = span.type;
```

Comment:
In the apm-server codebase, there is a special handling for inferring span.subtype and span.action from span.type if both subtype and action are not present in the input ([ref](https://github.com/elastic/apm-server/blob/d1b5503a66120b87949e34e82abfa2dc7e8cc9ea/model/modeldecoder/v2/decoder.go#L828-L837)).

For deducing `target_type` in this case, should the inferred value for `span.{type, subtype}` be used? Or can we remove the special handling of `span.{type, subtype}` since it seems to be added to keep backward compatibility?

Reply:
Points after discussion with Alex and Felix:

1. Need more context on the need to fallback to `span.{type, subtype}` for inferring `target.{type, name}`
2. For handling backward compatibility with `span.destination.service.resource`, do we need the conditional check for `span.type` to be one of `db`, `messaging` or `external`